File: src/lib/rankProjection.ts
/**
 * Final-rank projection.
 *
 * The bot reads an Estimated Rank ("C (3831)") each turn. This extrapolates the rating trajectory to the end of the career so
 * the run can show, mid-career, roughly what final grade it is on track for — and whether it will clear a target grade (e.g. the
 * grade needed to place well in a Champions Meeting). Rating only ever climbs, so the projection is floored at the current rating.
 *
 * Pure and dependency-free. The rating->grade mapping is supplied by the caller (the game's real RANK_MINS/RANK_LABELS table)
 * rather than duplicated here, so the projection stays a small, unit-testable numeric core.
 */

/** A per-turn rating observation. */
export interface RankObservation {
    turn: number
    rating: number
}

/** A grade band: the minimum rating required to reach `grade`. Sorted ascending by `minRating` by the consumer helpers. */
export interface GradeThreshold {
    grade: string
    minRating: number
}

export interface RankProjection {
    currentTurn: number
    currentRating: number
    currentGrade: string
    /** Projected rating at the end of the career (floored at the current rating). */
    projectedRating: number
    projectedGrade: string
    /** Fitted rating gained per turn (least-squares slope over the observations; 0 with fewer than two). */
    ratingPerTurn: number
}

/**
 * Resolves the grade for a rating: the highest grade whose minimum rating the value meets.
 *
 * @param rating The rating to classify.
 * @param thresholds Grade bands (any order; sorted internally).
 * @returns The grade label, or the lowest band's grade when the rating is below every threshold.
 */
export function gradeForRating(rating: number, thresholds: GradeThreshold[]): string {
    if (thresholds.length === 0) return ""
    const sorted = [...thresholds].sort((a, b) => a.minRating - b.minRating)
    let grade = sorted[0].grade
    for (const t of sorted) {
        if (rating >= t.minRating) grade = t.grade
        else break
    }
    return grade
}

/** Least-squares slope of rating over turn. Returns 0 when there are fewer than two points or the turns don't vary. */
function ratingSlope(obs: RankObservation[]): number {
    if (obs.length < 2) return 0
    const n = obs.length
    const meanT = obs.reduce((s, o) => s + o.turn, 0) / n
    const meanR = obs.reduce((s, o) => s + o.rating, 0) / n
    let cov = 0
    let varT = 0
    for (const o of obs) {
        cov += (o.turn - meanT) * (o.rating - meanR)
        varT += (o.turn - meanT) ** 2
    }
    return varT === 0 ? 0 : cov / varT
}

/**
 * Projects the final rank at the end of the career from the rating trajectory so far.
 *
 * @param observations Per-turn rating observations (at least one; more improve the slope estimate).
 * @param currentTurn The trainee's current turn.
 * @param totalTurns The final career turn to project to (e.g. 73 for a standard career).
 * @param thresholds The rating->grade table.
 * @returns The projection, or null when there are no observations.
 */
export function projectFinalRank(observations: RankObservation[], currentTurn: number, totalTurns: number, thresholds: GradeThreshold[]): RankProjection | null {
    if (observations.length === 0) return null

    // Current = the latest observation at or before the current turn (fall back to the last observation).
    const atOrBefore = observations.filter((o) => o.turn <= currentTurn)
    const current = (atOrBefore.length > 0 ? atOrBefore : observations).reduce((latest, o) => (o.turn >= latest.turn ? o : latest))

    const slope = ratingSlope(observations)
    const turnsLeft = Math.max(0, totalTurns - current.turn)
    const projectedRating = Math.max(current.rating, Math.round(current.rating + slope * turnsLeft))

    return {
        currentTurn,
        currentRating: current.rating,
        currentGrade: gradeForRating(current.rating, thresholds),
        projectedRating,
        projectedGrade: gradeForRating(projectedRating, thresholds),
        ratingPerTurn: slope,
    }
}

/**
 * Whether the projected final rank clears a target grade.
 *
 * @param projection A projection from [projectFinalRank].
 * @param targetGrade The grade to clear (must exist in `thresholds`).
 * @param thresholds The rating->grade table.
 * @returns True when the projected rating meets the target grade's minimum.
 */
export function isOnTrackForGrade(projection: RankProjection, targetGrade: string, thresholds: GradeThreshold[]): boolean {
    const target = thresholds.find((t) => t.grade === targetGrade)
    if (!target) return false
    return projection.projectedRating >= target.minRating
}

File: android/app/src/main/java/com/steve1316/uma_android_automation/bot/RankProjection.kt
package com.steve1316.uma_android_automation.bot

import com.steve1316.uma_scoring.scoreToRankLabel
import kotlin.math.pow

/**
 * A projected final rank, extrapolated from the per-turn Estimated Rank score trajectory.
 *
 * @property currentTurn The latest observed turn.
 * @property currentScore The latest observed Estimated Rank total score.
 * @property currentGrade The grade for [currentScore].
 * @property projectedScore The extrapolated total score at career end (floored at [currentScore] — rating only climbs).
 * @property projectedGrade The grade for [projectedScore].
 * @property scorePerTurn The fitted score gained per turn (least-squares slope).
 */
data class RankProjectionResult(
    val currentTurn: Int,
    val currentScore: Int,
    val currentGrade: String,
    val projectedScore: Int,
    val projectedGrade: String,
    val scorePerTurn: Double,
)

/**
 * Accumulates the per-turn Estimated Rank score (`estimateRank(...).totalScore`) during a run and extrapolates the final rank at
 * career end, so the run can show mid-career what grade it is on track for. The extrapolation is a least-squares fit floored at
 * the current score (rating only ever climbs). The pure fit helpers take explicit points so they are unit-testable; the object
 * holds the live run's observations and is reset at run start.
 */
object RankProjection {
    /** The career turn the projection extrapolates to (the finale). */
    const val DEFAULT_FINALE_TURN = 73

    /**
     * Recency half-life, in turns, for the weighted slope. A run's score trajectory is convex - it climbs slowly in Junior/early-Classic, then steeply in Senior as bonds max and
     * rainbows/spirit bursts land - so a flat least-squares fit anchored by the slow early turns badly under-projects mid-run. Weighting each observation by `2^(-turnsAgo/HALF_LIFE)`
     * makes the slope track the recent (steeper) trend, so the projection reflects where the run is actually heading. Points this many turns before the latest count half as much.
     */
    private const val RECENCY_HALF_LIFE_TURNS = 12.0

    private val observations = mutableListOf<Pair<Int, Int>>() // (turn, score), one per turn

    /** Clears the accumulated observations. Call at run start. */
    fun reset() = observations.clear()

    /** Records a turn's Estimated Rank score, replacing any prior observation for the same turn. */
    fun record(turn: Int, score: Int) {
        observations.removeAll { it.first == turn }
        observations.add(turn to score)
    }

    fun hasData(): Boolean = observations.isNotEmpty()

    /** Number of per-turn observations accumulated this run. Used to gate decisions (e.g. auto-abandon) that must not fire on a too-short trajectory, such as right after the bot is started mid-career. */
    fun observationCount(): Int = observations.size

    /** Projects the final rank at [totalTurns] from the live observations, or null when none have been recorded. */
    fun project(totalTurns: Int): RankProjectionResult? = projectFrom(observations, totalTurns)

    /**
     * Recency-weighted least-squares slope of score over turn, weighting each observation by `2^(-turnsAgo / RECENCY_HALF_LIFE_TURNS)` so recent turns dominate the fit and the
     * projection follows the run's accelerating trajectory instead of being dragged down by slow early turns. Reduces to the ordinary slope for a perfectly linear series (any
     * weighting of collinear points recovers the exact slope). Returns 0 with fewer than two points or when the turns do not vary.
     *
     * @param points (turn, score) observations.
     * @return Score gained per turn.
     */
    fun slope(points: List<Pair<Int, Int>>): Double {
        if (points.size < 2) return 0.0
        val latestTurn = points.maxOf { it.first }
        val weights = points.map { 2.0.pow(-(latestTurn - it.first).toDouble() / RECENCY_HALF_LIFE_TURNS) }
        val weightSum = weights.sum()
        var meanT = 0.0
        var meanS = 0.0
        for (i in points.indices) {
            meanT += weights[i] * points[i].first
            meanS += weights[i] * points[i].second
        }
        meanT /= weightSum
        meanS /= weightSum
        var cov = 0.0
        var varT = 0.0
        for (i in points.indices) {
            val dt = points[i].first - meanT
            cov += weights[i] * dt * (points[i].second - meanS)
            varT += weights[i] * dt * dt
        }
        return if (varT == 0.0) 0.0 else cov / varT
    }

    /**
     * Pure projection over an explicit observation list.
     *
     * @param points (turn, score) observations (need at least one).
     * @param totalTurns The final career turn to project to.
     * @return The projection, or null when [points] is empty.
     */
    fun projectFrom(points: List<Pair<Int, Int>>, totalTurns: Int): RankProjectionResult? {
        if (points.isEmpty()) return null
        val current = points.maxByOrNull { it.first }!!
        val m = slope(points)
        val turnsLeft = maxOf(0, totalTurns - current.first)
        val projectedScore = maxOf(current.second, Math.round(current.second + m * turnsLeft).toInt())
        return RankProjectionResult(
            currentTurn = current.first,
            currentScore = current.second,
            currentGrade = scoreToRankLabel(current.second),
            projectedScore = projectedScore,
            projectedGrade = scoreToRankLabel(projectedScore),
            scorePerTurn = m,
        )
    }
}
